### PR TITLE
JSON input

### PR DESCRIPTION
**Pull request by stigrj**
Preparing for the new input parser
------------------------------------------

- [x] Add `nlohmann/json` library as external project
- [x] Rewrite `mrenv` input using JSON
- [x] Rewrite `SCFDriver` input using JSON
- [x] Add `parselglossy` input parser
- [x] Remove `getkw`
- [x] Sayonara `boost`

The input is now separated into `user_input` and `program_input`, with different definitions of sections and keywords. The `mrchem` parser script runs `parselglossy` on the `user_input` which translates from the good old `getkw` syntax to JSON and validates the parameters. Then the validated `user_input` is translated into `program_input` which is then passed to the `mrchem.x` program.

At this point the `user_input` is identical to the old input structure, and I have written a translator into the new `program_input`, which means that old input files should still work (at least for the features that are integration tested), but the `user_input` structure will soon be revised and replaced.

### CODE REVIEW

**Comment by stigrj** [comment]
@robertodr the `-DJSON_BuildTests=OFF` option in `nlohmann_json` won't take. Any idea why?

**Comment by stigrj** [comment]
Actually, it seems none of the `CMAKE_ARGS` in the the `FetchContent_Populate` blocks are used, all options end up with their default values regardless, also in MRCPP. The only way to overwrite is through `--cmake-options="-DJSON_BuildTests=OFF"` etc. in the MRChem `./setup`.

**Comment by stigrj** [comment]
Is this perhaps the way to go: `set(JSON_BuildTests OFF CACHE BOOL "" FORCE)`?
Taken from here: https://github.com/dev-cafe/cmake-cookbook/blob/master/chapter-04/recipe-03/cxx-example/CMakeLists.txt

**Comment by robertodr** [comment]
Yep, that solves it. I've run into this from time to time and I suspect it's a bug in `FetchContent`. I don't have time to investigate it and open an issue on CMake's repo though. Paging @bast who might know more.

**Comment by robertodr** [comment]
BTW, there is an isolated repo that shows how to pull the C++ JSON library with both superbuild styles: https://github.com/dev-cafe/cpp-parselglossy/tree/master
This commit shows the problem: https://github.com/dev-cafe/cpp-parselglossy/commit/9ea5a5cc51a6790b48b99c8c9a6abe51996b005f
This commit has the abovementioned fix: https://github.com/dev-cafe/cpp-parselglossy/commit/abe3d0ca89f9e831a242e1bdfeee7e82d8759274

File: cmake/custom/main.cmake [hunk 1]
```diff
@@ -6,7 +6,6 @@ configure_file (
 set(PYTHON_SITE_INSTALL_DIR ${CMAKE_INSTALL_LIBDIR}/python${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}/site-packages)
 
 # Fetch dependencies
-include(${PROJECT_SOURCE_DIR}/external/upstream/fetch_getkw.cmake)
```
Comment (robertodr) [inline cmake/custom/main.cmake]: I did not hear the :champagne: pop. :smile_cat:

Comment (bast) [reply cmake/custom/main.cmake]: Nice!

Comment (stigrj) [reply cmake/custom/main.cmake]: Yeah, it's not *really* working yet :slightly_smiling_face:

Comment (robertodr) [reply cmake/custom/main.cmake]: So :champagne:?

File: src/input/template.yml [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  - name: est_norm
+    type: float
+    default: -1.0
+    docstring: |
+      Norm used in translation from abs_prec to rel_prec.
+  - name: energy_unit
+    type: str
+    default: 'au'
+    predicates:
+      - "value == 'au' or value == 'kcal' or value == 'kJ'"
```
Comment (robertodr) [inline src/input/template.yml]: `value in ['au', 'kcal', 'kJ']`

File: src/driver.cpp [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+
+    return success;
+}
+
+void driver::calc_scf_properties(const json &json_prop, Molecule &mol) {
+    auto &nuclei = mol.getNuclei();
+    auto &Phi = mol.getOrbitals();
+
+    auto json_dipole = json_prop.find("dipole_moment");
+    if (json_dipole != json_prop.end()) {
```
Comment (robertodr) [inline src/driver.cpp]: Does this check ever fail? The final representation of the input is complete in the sense that it has all fields from the template. That is, this section will always be present. I think you'll need to insert a list keyword in the parent section to give a list of properties to calculate. Then at this level you'd iterate over the list and dispatch to constuctors.

File: src/mrchem.in [hunk 4]
```diff
@@ -25,103 +25,56 @@
...
+
+    if not dryrun:
+        cmd = executable + ' ' + xfile
+        p = subprocess.call(cmd, shell=True)
 
 def parse_cmdline():
-    usage = "usage: %prog [options] [config] inpfile"
-    version = "@PROGRAM_VERSION@"
+    usage = "usage: %prog [options] [config] inp_file"
+    version = "0.0.1"
```
Comment (robertodr) [inline src/mrchem.in]: I think this is unwanted.

File: src/mrchem.in [hunk 5]
```diff
@@ -145,7 +98,7 @@ def parse_cmdline():
         action='store',
         dest='executable',
         type='string',
-        default='@MRCHEM_EXECUTABLE@',
+        default='/home/stig/src/mrchem-stig/install-json/bin/mrchem.x',
```
Comment (robertodr) [inline src/mrchem.in]: Same here. You might have checked in the wrong version of this file

File: src/mrchem.in [hunk 6]
```diff
@@ -25,103 +25,56 @@
...
+    executable, dryrun, printlevel, inp_file_cmd = parse_cmdline()
+
+    in_path = Path(inp_file_cmd)
+    tmpl_path = Path('@PROJECT_SOURCE_DIR@/src/input/template.yml')
+
+    # read user input file (JSONDict <- getkw file)
+    user_dict = parse(infile=in_path, template=tmpl_path, grammar='getkw')
+
+    # run user defined verification (JSONDict <- JSONDict)
+    verify_prec(user_dict)
```
Comment (robertodr) [inline src/mrchem.in]: Can't these checks on precision be absorbed into the `template.yml`? Is the format too rigid?

File: src/mrchem.in [hunk 7]
```diff
@@ -25,103 +25,56 @@
...
 
+    # now that all keywords have sensible values,
+    # we can translate user input into program input
+    program_dict = translate_input(user_dict)
```
Comment (robertodr) [inline src/mrchem.in]: Why is this needed?

File: src/mrchem.in [hunk 8]
```diff
@@ -158,455 +111,418 @@ def parse_cmdline():
...
+
+def write_printer(program_dict, user_dict):
+  program_dict["printer"] = {
+    "filename": "mrchem",
+    "printprec": 15,
+    "printlevel": user_dict["printlevel"],
+    "teletype": user_dict["teletype"]
+  }
+
+def write_molecule(program_dict, user_dict):
```
Comment (robertodr) [inline src/mrchem.in]: This is clearly a case where having the ability to perform an intermediate action between defaulting and predicate checking might be helpful. Duly noted for next iteration on `parselglossy`.

File: src/mrchem.in (hunk 5 above)
Comment (stigrj) [reply src/mrchem.in]: Woops!

File: src/mrchem.in (hunk 6 above)
Comment (stigrj) [reply src/mrchem.in]: The logic is a bit too complicated, I think

File: src/input/template.yml (hunk 2 above)
Comment (stigrj) [reply src/input/template.yml]: Thanks

File: src/mrchem.in (hunk 7 above)
Comment (stigrj) [reply src/mrchem.in]: The problem is that the `program_input` is very different from the `user_input`, and I don't want to expose all the detailed implementation specific keyword options to the user.

The idea is to have a more familiar and user friendly `user_input` which is verified through parselglossy, and then use this script to "build" a complete calculation from the `user_input`, where the resulting `program_input` defines all the steps in the calculation in the sense that *if* a given section is present *then* this calculation will take place.

This also goes for your question above: if "dipole_moment" is present in the `program_input`, then the dipole moment is computed (a list as you suggest might still be a better option for the properties, though).

File: Pipfile [hunk 9]
```diff
@@ -4,6 +4,7 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
+parselglossy = "==0.3.0"
 pyparsing = "*"
```
Comment (robertodr) [inline Pipfile]: You can remove `pyparsing`, it will be fetched as part of the dependencies of `parselglossy`.

File: src/input/template.html [hunk 10]
```diff
@@ -0,0 +1,1005 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (robertodr) [inline src/input/template.html]: I believe this file was accidentally checked in?

File: src/input/template.rst [hunk 11]
```diff
@@ -0,0 +1,549 @@
+.. raw:: html
```
Comment (robertodr) [inline src/input/template.rst]: This file should be moved under the `doc` folder and included somewhere in `mrchem_manual.rst`

File: src/input/template.html (hunk 10 above)
Comment (stigrj) [reply src/input/template.html]: Damnation
